lib/framedivider.rb

```diff
@@ -390,46 +422,284 @@ def opening(model, id)
+    #     /   \
+    #    /     \
+    #   / C --- D    <<< allowed as OpenStudio/E+ subsurface?
+    #  / /
+    #  B
+    #
+    # Should convert (annoying) 4-point subsurface into triangle ...
+    #        A
+    #       / \
+    #      /   \
+    #     /     \
+    #    B - C - D   <<< allowed as OpenStudio/E+ subsurface?
+    #
+    four = true if s.vertices.size == 4
+
+    if s.windowPropertyFrameAndDivider.empty?
+      vec = s.vertices
+      area = gross
+    else
+      width = s.windowPropertyFrameAndDivider.get.frameWidth
+      ptz = offset(s.vertices, width)
+      fd = true
+
+      # Re-convert Topolys 3D points into OpenStudio 3D points.
+      vec = OpenStudio::Point3dVector.new
+      vec << OpenStudio::Point3d.new(ptz[:A][:p].x, ptz[:A][:p].y, ptz[:A][:p].z)
+      vec << OpenStudio::Point3d.new(ptz[:B][:p].x, ptz[:B][:p].y, ptz[:B][:p].z)
+      vec << OpenStudio::Point3d.new(ptz[:C][:p].x, ptz[:C][:p].y, ptz[:C][:p].z)
+      vec << OpenStudio::Point3d.new(ptz[:D][:p].x, ptz[:D][:p].y, ptz[:D][:p].z) if four
+
+      tr1 = []
+      tr1 << ptz[:A][:p]
+      tr1 << ptz[:B][:p]
+      tr1 << ptz[:C][:p]
+      area1 = areaHeron(tr1)
+
+      area2 = 0
+      if four
+        tr2 = []
+        tr2 << ptz[:A][:p]
+        tr2 << ptz[:C][:p]
+        tr2 << ptz[:D][:p]
+        area2 = areaHeron(tr2)
+      end
+      area = area1 + area2
+    end
+
+    sub = { v:      s.vertices,
+            points: vec,
+            n:      n,
+            gross:  gross,
+            area:   area,
+            type:   type,
+            u:      u }
+
+    sub[:glazed] = true if glazed
+    subs[id] = sub
+  end
+
+  # Test for conflicts (with fits?, overlaps?) between surfaces to determine
+  # whether to keep original points or switch to std::vector of revised
+  # coordinates, offset by Frame & Divider frame width. This will also
+  # inadvertently catch pre-existing (yet nonetheless invalid) OpenStudio inputs
+  # (without Frame & Dividers).
+  valid = true
+  subs.each do |id, sub|
+    next unless fd
+    next unless valid
+    unless fits?(surface.vertices, sub[:points])
+      msg = "Subsurface '#{id}' can't fit in '#{identifier}' - skipping"
+      TBD.log(TBD::ERROR, msg)
+      valid = false
+    end
+    subs.each do |i, sb|
+      next unless valid
+      next if i == id
+      if overlaps?(sub[:points], sb[:points])
+        msg = "Subsurface '#{id}' overlaps sibling '#{i}' - skipping"
+        TBD.log(TBD::ERROR, msg)
+        valid = false
+      end
+    end
+  end
+
+  if fd
+    if valid
+      # No conflicts. Reset subsurface gross area.
+      subs.values.each { |sub| sub[:gross] = sub[:area] }
+    else
+      # One or more conflicts between the parent surface & one or more
+      # subsurfaces, or between subsurfaces. Ignore Frame & Divider offsets
+      # and revert to original vertices.
+      subs.values.each { |sub| sub[:points] = sub[:v] }
+      subs.values.each { |sub| sub[:area] = sub[:gross] }
```

Comment:
Want to log this right?

Reply:
It does - in the previous [loop](https://github.com/rd2/tbd/blob/86bde630e3331f133f55097e25017cea2e11f96f/lib/framedivider.rb#L652).